man/cdpth2tga.1
.Dd July 23, 2015
.Dt CDPTH2TGA 1
.Os
.Sh NAME
.Nm cdpth2tga
.Nd CDPTH depth image to TGA converter
.Sh SYNOPSIS
.Nm cdpth2tga
.Op Ar options
.Ar cdpth 2da tga
.Sh DESCRIPTION
.Nm
converts CDPTH depth information files into common TGA images.
CDPTH, Compressed DePTH, are used as depth information for area
background images in BioWare's Nintendo DS game
.Em Sonic Chronicles: The Dark Brotherhood .
.Pp
Layout-wise, a CDPTH is stored similar to CBGT:
cells of 64\(mu64 pixels, compressed using Nintendo's 0x10 LZSS algorithm.
Unlike CBGT, though, the cells themselves are
.Em not
swizzled into 8\(mu8
tiles, and the pixel value in CDPTH is a 16bit integer specifying
a depth.
.Pp
The width and height of the final image is not stored within the
CDPTH file, and has to be provided from the outside, like from
the dimensions of the CBGT image, the dimensions of the 2DA file
for the CBGT image.
.Pp
The depth information is stored as 16-bit values, but only the
lower 15 bits are used.
For conversion into a grayscale image, those 15 bits of grayscale
information are further compressed to 8-bit.
.Sh OPTIONS
.Bl -tag -width xxxx -compact
.It Fl h
.It Fl Fl help
Show a help text and exit.
.It Fl Fl version
Show version information and exit.
.El
.Bl -tag -width xx -compact
.It Ar cdpth
The name of the CDPTH file to read.
The CDPTH contains the depth information part of the image.
.It Ar 2da
The name of the 2DA file to use.
The 2DA is necessary to know how many cells are in the CDPTH file.
.It Ar tga
The resulting TGA file will be written there.
.El
.Sh EXAMPLE
Convert a CDPTH+2DA into a TGA:
.Pp
.Dl $ cdpth2tga a.cdpth a.2da result.tga
.Sh SEE ALSO
.Xr cbgt2tga 1
.Pp
More information about the xoreos project can be found on
.Lk https://xoreos.org/ "its website"
.Ns .
.Sh AUTHORS
This program is part of the xoreos-tools package, which in turn is
part of the xoreos project, and was written by the xoreos team.
Please see the
.Pa AUTHORS
file for details.


NEWS
Changes in xoreos-tools version 0.0.2
=====================================

This is the first official release of xoreos-tools, together with xoreos.

Included are:

- Tools for handling resource archives
  - unkeybif
  - unerf
  - unrim
  - unherf
  - unnds
  - desmall
- Tools converting basic resource formats into human-readable form
  - gff2xml
  - tlk2xl
  - convert2da
- Tools converting graphics formats into TGA
  - xoreostex2tga
  - unnsbtx
  - nbfs2tga
  - ncgr2tga
  - cbgt2tga
  - cdpth2tga


src/cdpth2tga.cpp
/* xoreos-tools - Tools to help with xoreos development
 *
 * xoreos-tools is the legal property of its developers, whose names
 * can be found in the AUTHORS file distributed with this source
 * distribution.
 *
 * xoreos-tools is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * xoreos-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with xoreos-tools. If not, see <http://www.gnu.org/licenses/>.
 */

/** @file
 *  Tool to convert CDPTH depth images to TGA. */

#include <cstring>
#include <cstdio>
#include <cstdlib>

#include "src/version/version.h"

#include "src/common/ustring.h"
#include "src/common/util.h"
#include "src/common/strutil.h"
#include "src/common/error.h"
#include "src/common/platform.h"
#include "src/common/readfile.h"
#include "src/common/cli.h"

#include "src/aurora/types.h"
#include "src/aurora/util.h"
#include "src/aurora/2dafile.h"

#include "src/images/cdpth.h"

#include "src/util.h"

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Common::UString &cdpthFile, Common::UString &twoDAFile,
                      Common::UString &outFile);

void convert(const Common::UString &cdpthFile, const Common::UString &twoDAFile,
             const Common::UString &outFile);

int main(int argc, char **argv) {
	initPlatform();

	try {
		std::vector<Common::UString> args;
		Common::Platform::getParameters(argc, argv, args);

		int returnValue = 1;
		Common::UString cdpthFile, twoDAFile, outFile;

		if (!parseCommandLine(args, returnValue, cdpthFile, twoDAFile, outFile))
			return returnValue;

		convert(cdpthFile, twoDAFile, outFile);
	} catch (...) {
		Common::exceptionDispatcherError();
	}

	return 0;
}

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Common::UString &cdpthFile, Common::UString &twoDAFile,
                      Common::UString &outFile) {
	using Common::CLI::Parser;
	using Common::CLI::ValGetter;
	using Common::CLI::NoOption;
	using Common::CLI::makeEndArgs;
	std::vector<Common::CLI::Getter *> getters;
	NoOption cdpthFileOpt(false, new ValGetter<Common::UString &>(cdpthFile, "cdpth"));
	NoOption twoDAFileOpt(false, new ValGetter<Common::UString &>(twoDAFile, "2da"));
	NoOption outFileOpt(false, new ValGetter<Common::UString &>(outFile, "tga"));
	Parser parser(argv[0], "CDPTH depth image to TGA converter", "", returnValue,
	              makeEndArgs(&cdpthFileOpt, &twoDAFileOpt, &outFileOpt));

	return parser.process(argv);
}

static void getDimensions(const Common::UString &twoDAFile, uint32 &width, uint32 &height) {

	Common::ReadFile twoDAStream(twoDAFile);
	Aurora::TwoDAFile twoDA(twoDAStream);

	width  = twoDA.getColumnCount() * 64;
	height = twoDA.getRowCount()    * 64;
}

void convert(const Common::UString &cdpthFile, const Common::UString &twoDAFile,
             const Common::UString &outFile) {

	uint32 width, height;
	getDimensions(twoDAFile, width, height);

	Common::ReadFile cdpth(cdpthFile);
	Images::CDPTH image(cdpth, width, height);

	image.flipVertically();

	image.dumpTGA(outFile);
}
